### DISCUSSION TITLE
Help setting up NSF model

### DISCUSSION THREAD

**Discussion by austin-hoover** [mathisgerdes/bijx#4] [Q&A] [answered]
Hi there,

I'm wanting to train a normalizing flow built from RQS coupling layers. Here's my attempt for a 2D problem based on the documentation, along with a comparison to zuko (which is using autoregressive instead of coupling layers, but performance should be comparable). 
[test_bijx.py](https://github.com/user-attachments/files/26358045/test_bijx.py)
[test_zuko.py](https://github.com/user-attachments/files/26358046/test_zuko.py)

Each layer uses an MLP of width 64 and depth 3, and there are three total layers. Here are the training results for bijx and then for zuko after 1000 iterations of the ADAM optimizer:
<img width="400" alt="fig_bijx"   src="https://github.com/user-attachments/assets/ad128e16-e5fb-47a7-a9cc-67dc05955ee0" />
<img width="400" alt="fig_zuko" src="https://github.com/user-attachments/assets/eab83e6f-94b8-4bca-9607-f95c6eee1236" />

I wondered if you could help diagnose what's going wrong when building or training the NSF model in bijx.

Thanks,
Austin

**Comment by mathisgerdes** [comment]
Hi Austin,

My first intuition is try to reduce the initialization scale or suppress the final outputs in the coupling flow. Try one of:

- Add a warm up phase, slowly increasing the learning rate
- Check the final kernel / bias parameters in the MLPs and reduce their initialization scale
- Divide the output by some factor 

The top right picture looks like the bijections may have been pushed into "extreme" regimes. 

Let me know if that works, I can look into it more in a bit...

**Comment by austin-hoover** [comment]
Thanks for the fast response. I'll take a look at the MLP initialization.

**Comment by mathisgerdes** [comment, marked as answer]
I've had another look, and there are a few issues with your example script that are actually much simpler:

1. You call the numpy function `sample_data` inside the jit-compiled update loop. The way that Jax works, this means training only ever sees one batch (not multiple different random seeds).
2. Spline by default actively transforms the fixed interval [0, 1], which is too small for your data. This can be fixed by manually rescaling the data before/after the spline bijection, or adding learnable scalings.

The simplest fix for the first issue is just to pass new samples `x` as one input to `update_step(model, optimizer, x)`, then `nnx.value_and_grad(loss_fn)(model, x)` with added argument `x`.

For learnable scaling, you could use
```
spline = bijx.MonotoneRQSpline(knots, (), rngs=rngs)
spline = bijx.Chain(bijx.AffineLinear(rngs=rngs), spline, bijx.AffineLinear(rngs=rngs))
spline_template = bijx.ModuleReconstructor(spline)
```

Or to fix the active bounds of the spline to -3 and 3 (which I think NSF uses, but you should double check):
```
spline = bijx.MonotoneRQSpline(knots, (), rngs=rngs)
rescale = bijx.AffineLinear(scale=bijx.Const(6.0), shift=bijx.Const(-3.0), transform_scale=None)
spline = bijx.Chain(rescale.invert(), spline, rescale)
spline_template = bijx.ModuleReconstructor(spline)
```

With the latter I get the more reasonable result below:
<img width="511" height="261" alt="image" src="https://github.com/user-attachments/assets/cd6b089d-85f8-477b-bac2-372d58211db5" />

Tuning MLP, initializations, and other hyperparameters can probably improve things further. Hope that helps!

**Reply by mathisgerdes** [reply]
And as a final piece of unsolicited advertisement for other work: the issue of splines having these fixed "active intervals" led us to search for [analytic bijections that are smooth and act on all of R](https://arxiv.org/abs/2601.10774). They can be used here as a drop-in replacement like this:

```
n_copies = 16  # increase / decrease to change expressivity, similar to knot count
@nnx.split_rngs(splits=n_copies)
@nnx.vmap
def get_bijection(rngs):
    return bijx.ScanChain(bijx.CubicRational(rngs=rngs))

template = bijx.ModuleReconstructor(get_bijection(rngs))
```

**Reply by austin-hoover** [reply]
That makes sense on the sampling function and on the spline intervals. The default boundary in zuko was [-5, 5]. I was originally interested in testing these analytic bijections to get rid of this boundary, but wanted to start from the splines for comparison. I'm also new to jax. Thanks a lot!
